Firestore/core/src/core/sync_engine.cc

```diff
@@ -567,6 +573,74 @@ void SyncEngine::RemoveLimboTarget(const DocumentKey& key) {
   PumpEnqueuedLimboResolutions();
 }
 
+void SyncEngine::LoadBundle(std::shared_ptr<bundle::BundleReader> reader,
+                            std::shared_ptr<api::LoadBundleTask> result_task) {
+  auto bundle_metadata = reader->GetBundleMetadata();
+  if (!reader->reader_status().ok()) {
+    LOG_WARN("Failed to GetBundleMetadata() for bundle with error %s",
+             reader->reader_status().error_message());
+    result_task->SetError(reader->reader_status());
+    return;
+  }
+
+  auto has_newer_bundle = local_store_->HasNewerBundle(bundle_metadata);
+  if (has_newer_bundle) {
+    result_task->SetSuccess(SuccessProgress(bundle_metadata));
+    // TODO: reader.close()?
+    return;
+  }
+
+  result_task->UpdateProgress(InitialProgress(bundle_metadata));
+
+  BundleLoader loader(local_store_, bundle_metadata);
+
+  int64_t current_bytes_read = 0;
+  // Breaks when either error happened, or when there is no more element to
+  // read.
+  while (true) {
+    auto element = reader->GetNextElement();
+    if (!reader->reader_status().ok()) {
+      LOG_WARN("Failed to GetNextElement() from bundle with error %s",
+               reader->reader_status().error_message());
+      result_task->SetError(reader->reader_status());
+      return;
+    }
+
+    // No more elements from reader.
+    if (element == nullptr) {
+      break;
+    }
+
+    int64_t old_bytes_read = current_bytes_read;
+    current_bytes_read = reader->bytes_read();
+    auto progress = loader.AddElement(std::move(element),
```

Comment:
Optional: consider naming this variable so that it's more obvious that it is a `StatusOr` (e.g. `maybe_progress`).